service/history/mutableStateBuilder.go

```diff
@@ -2318,42 +2343,80 @@ func (e *mutableStateBuilder) ReplicateWorkflowExecutionContinuedAsNewEvent(sour
-		DecisionTimeoutValue:        newExecutionInfo.DecisionTimeoutValue,
-		ExecutionContext:            nil,
-		NextEventID:                 newStateBuilder.GetNextEventID(),
-		LastProcessedEvent:          common.EmptyEventID,
-		TransferTasks:               newTransferTasks,
-		DecisionVersion:             di.Version,
-		DecisionScheduleID:          di.ScheduleID,
-		DecisionStartedID:           di.StartedID,
-		DecisionStartToCloseTimeout: di.DecisionTimeout,
-		ContinueAsNew:               true,
-		PreviousRunID:               prevRunID,
-		ReplicationState:            newStateBuilder.GetReplicationState(),
+		RequestID:            uuid.New(),
+		DomainID:             domainID,
+		Execution:            newExecution,
+		ParentDomainID:       parentDomainID,
+		ParentExecution:      parentExecution,
+		InitiatedID:          initiatedID,
+		TaskList:             newExecutionInfo.TaskList,
+		WorkflowTypeName:     newExecutionInfo.WorkflowTypeName,
+		WorkflowTimeout:      newExecutionInfo.WorkflowTimeout,
+		DecisionTimeoutValue: newExecutionInfo.DecisionTimeoutValue,
+		ExecutionContext:     nil,
+		NextEventID:          newStateBuilder.GetNextEventID(),
+		LastProcessedEvent:   common.EmptyEventID,
+		ContinueAsNew:        true,
+		PreviousRunID:        prevRunID,
+		ReplicationState:     newStateBuilder.GetReplicationState(),
+		HasRetryPolicy:       startedAttributes.RetryPolicy != nil,
+		Attempt:              e.executionInfo.Attempt,
+		InitialInterval:      e.executionInfo.InitialInterval,
+		BackoffCoefficient:   e.executionInfo.BackoffCoefficient,
+		MaximumInterval:      e.executionInfo.MaximumInterval,
+		ExpirationTime:       e.executionInfo.ExpirationTime,
+		MaximumAttempts:      e.executionInfo.MaximumAttempts,
+		NonRetriableErrors:   e.executionInfo.NonRetriableErrors,
+	}
+	if continueAsNewAttributes.GetBackoffStartIntervalInSeconds() > 0 {
+		// this is a retry
+		continueAsNew.Attempt++
+	}
+
+	// timeout includes workflow_timeout + backoff_interval
+	timeoutInSeconds := continueAsNewAttributes.GetExecutionStartToCloseTimeoutSeconds() + continueAsNewAttributes.GetBackoffStartIntervalInSeconds()
+	timeoutDuration := time.Duration(timeoutInSeconds) * time.Second
+	timeoutDeadline := time.Now().Add(timeoutDuration)
+	if !e.executionInfo.ExpirationTime.IsZero() && timeoutDeadline.After(e.executionInfo.ExpirationTime) {
+		// expire before timeout
+		timeoutDeadline = e.executionInfo.ExpirationTime
+	}
+	continueAsNew.TimerTasks = []persistence.Task{&persistence.WorkflowTimeoutTask{
+		VisibilityTimestamp: timeoutDeadline,
+	}}
+
+	if di != nil {
+		continueAsNew.DecisionVersion = di.Version
+		continueAsNew.DecisionScheduleID = di.ScheduleID
+		continueAsNew.DecisionStartedID = di.StartedID
+		continueAsNew.DecisionStartToCloseTimeout = di.DecisionTimeout
+
+		if newStateBuilder.GetReplicationState() != nil {
+			newStateBuilder.UpdateReplicationStateLastEventID(sourceClusterName, startedEvent.GetVersion(), di.ScheduleID)
```

Comment:
We need to update replication state in both cases.  Regular case should use the di.ScheduleID and retry case should still set it to startedEventID.

Reply:
added